src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -88,56 +90,64 @@ end
+    return Adjoint(A), Adjoint_mat_pullback
 end
 
+_Adjoint_vec_pullback(ȳ::Tangent) = (NoTangent(), vec(ȳ.parent))
+_Adjoint_vec_pullback(ȳ::AbstractMatrix) = (NoTangent(), vec(adjoint(ȳ)))
+_Adjoint_vec_pullback(ȳ::AbstractThunk) = return _Adjoint_vec_pullback(unthunk(ȳ))
 function rrule(::Type{<:Adjoint}, A::AbstractVector{<:Number})
-    Adjoint_pullback(ȳ::Tangent) = (NoTangent(), vec(ȳ.parent))
-    Adjoint_pullback(ȳ::AbstractMatrix) = (NoTangent(), vec(adjoint(ȳ)))
-    return Adjoint(A), Adjoint_pullback
+    return Adjoint(A), _Adjoint_vec_pullback
 end
 
+_adjoint_mat_pullback(ȳ::Tangent) = (NoTangent(), ȳ.parent)
+_adjoint_mat_pullback(ȳ::AbstractVecOrMat) = (NoTangent(), adjoint(ȳ))
+_adjoint_mat_pullback(ȳ::AbstractThunk) = return _adjoint_mat_pullback(unthunk(ȳ))
 function rrule(::typeof(adjoint), A::AbstractMatrix{<:Number})
-    adjoint_pullback(ȳ::Tangent) = (NoTangent(), ȳ.parent)
-    adjoint_pullback(ȳ::AbstractVecOrMat) = (NoTangent(), adjoint(ȳ))
-    return adjoint(A), adjoint_pullback
+    return adjoint(A), _adjoint_mat_pullback
 end
 
+_adjoint_vec_pullback(ȳ::Tangent) = (NoTangent(), vec(ȳ.parent))
+_adjoint_vec_pullback(ȳ::AbstractMatrix) = (NoTangent(), vec(adjoint(ȳ)))
+_adjoint_vec_pullback(ȳ::AbstractThunk) = return _adjoint_vec_pullback(unthunk(ȳ))
 function rrule(::typeof(adjoint), A::AbstractVector{<:Number})
-    adjoint_pullback(ȳ::Tangent) = (NoTangent(), vec(ȳ.parent))
-    adjoint_pullback(ȳ::AbstractMatrix) = (NoTangent(), vec(adjoint(ȳ)))
-    return adjoint(A), adjoint_pullback
+    return adjoint(A), _adjoint_vec_pullback
 end
 
 #####
 ##### `Transpose`
 #####
 
+_Transpose_mat_pullback(ȳ::Tangent) = (NoTangent(), ȳ.parent)
+_Transpose_mat_pullback(ȳ::AbstractVecOrMat) = (NoTangent(), Transpose(ȳ))
+_Transpose_mat_pullback(ȳ::AbstractThunk) = return _Transpose_mat_pullback(unthunk(ȳ))
 function rrule(::Type{<:Transpose}, A::AbstractMatrix{<:Number})
-    Transpose_pullback(ȳ::Tangent) = (NoTangent(), ȳ.parent)
-    Transpose_pullback(ȳ::AbstractVecOrMat) = (NoTangent(), Transpose(ȳ))
-    return Transpose(A), Transpose_pullback
+    return Transpose(A), _Transpose_mat_pullback
 end
 
+_Transpose_vec_pullback(ȳ::Tangent) = (NoTangent(), vec(ȳ.parent))
+_Transpose_vec_pullback(ȳ::AbstractMatrix) = (NoTangent(), vec(Transpose(ȳ)))
+_Transpose_vec_pullback(ȳ::AbstractThunk) = return _Transpose_vec_pullback(unthunk(ȳ))
 function rrule(::Type{<:Transpose}, A::AbstractVector{<:Number})
-    Transpose_pullback(ȳ::Tangent) = (NoTangent(), vec(ȳ.parent))
-    Transpose_pullback(ȳ::AbstractMatrix) = (NoTangent(), vec(Transpose(ȳ)))
-    return Transpose(A), Transpose_pullback
+    return Transpose(A), _Transpose_vec_pullback
 end
 
+_transpose_mat_pullback(ȳ::Tangent) = (NoTangent(), ȳ.parent)
+_transpose_mat_pullback(ȳ::AbstractVecOrMat) = (NoTangent(), transpose(ȳ))
+_transpose_mat_pullback(ȳ::AbstractThunk) = return _transpose_mat_pullback(unthunk(ȳ))
```

Comment:
If these functions are define inside the rrule, type inference breaks. 🤷

Reply:
should add a comment to this effect